File: python/flatten_messy_array.py
def collect_lowest_level_arrays(arr):
    result = []
    for element in arr:
        if isinstance(element, list):
            result.extend(collect_lowest_level_arrays(element))
    return [arr] if not result else result

def near_flatten(array):
    arrays = collect_lowest_level_arrays(array)
    return sorted(arrays, key=lambda sub_array: min(sub_array))

File: javascript/flattenMessyArray.js
function collectLowestLevelArrays(arr) {
    let result = [];
    for (const x of arr) {
      if (Array.isArray(x)) {
        result = result.concat(collectLowestLevelArrays(x));
      }
    }
    return result.length === 0 ? [arr] : result;
  }
  
  function nearFlatten(array) {
    const lowestLevelArrays = collectLowestLevelArrays(array);
    return lowestLevelArrays.sort((a, b) => Math.min(...a) - Math.min(...b));
  }